clang/lib/CConv/ProgramInfo.cpp

```diff
@@ -763,76 +755,84 @@ bool ProgramInfo::computeInterimConstraintState
   // Get all the valid vars of interest i.e., all the Vars that are present
   // in one of the files being compiled.
   CAtoms ValidVarsVec;
+  std::set<Atom *> AllValidVars;
   for (const auto &I : Variables) {
     std::string FileName = I.first.getFileName();
-    if (FilePaths.count(FileName)) {
-      ConstraintVariable *C = I.second;
-      if (C->isForValidDecl()) {
-        CAtoms tmp = getVarsFromConstraint(C);
-        ValidVarsVec.insert(ValidVarsVec.begin(), tmp.begin(), tmp.end());
-      }
+    ConstraintVariable *C = I.second;
+    if (C->isForValidDecl()) {
+      CAtoms Tmp;
+      getVarsFromConstraint(C, Tmp);
+      AllValidVars.insert(Tmp.begin(), Tmp.end());
+      if (FilePaths.count(FileName) ||
+          FileName.find(BaseDir) != std::string::npos)
+        ValidVarsVec.insert(ValidVarsVec.begin(), Tmp.begin(), Tmp.end());
     }
   }
+
   // Make that into set, for efficiency.
   std::set<Atom *> ValidVarsS;
-  CVars ValidVarsKey;
   ValidVarsS.insert(ValidVarsVec.begin(), ValidVarsVec.end());
 
-  std::transform(ValidVarsS.begin() , ValidVarsS.end(),
-                 std::inserter(ValidVarsKey, ValidVarsKey.end()) ,
-                 [](const Atom *val){
-    if (const VarAtom *VA = dyn_cast<VarAtom>(val)) {
+  auto GetLocOrZero = [](const Atom *val) {
+    if (const auto *VA = dyn_cast<VarAtom>(val))
       return VA->getLoc();
-    }
-    return (uint32_t)0;
-  });
+    return (ConstraintKey) 0;
+  };
+  CVars ValidVarsKey;
+  std::transform(ValidVarsS.begin(), ValidVarsS.end(),
+                 std::inserter(ValidVarsKey, ValidVarsKey.end()), GetLocOrZero);
+  CVars AllValidVarsKey;
+  std::transform(AllValidVars.begin(), AllValidVars.end(),
+                 std::inserter(AllValidVarsKey, AllValidVarsKey.end()),
+                 GetLocOrZero);
 
   CState.Clear();
-
-  auto &RCMap = CState.RCMap;
-  auto &SrcWMap = CState.SrcWMap;
-  auto &TotalNDirectWPtrs = CState.TotalNonDirectWildPointers;
-  auto &InSrInDirectWPtrs = CState.InSrcNonDirectWildPointers;
-  CVars &WildPtrs = CState.AllWildPtrs;
-  CVars &InSrcW = CState.InSrcWildPtrs;
-  WildPtrs.clear();
   std::set<Atom *> DirectWildVarAtoms;
-  std::set<Atom *> IndirectWildAtoms;
-  auto &ChkCG = CS.getChkCG();
-  ChkCG.getSuccessors(CS.getWild(), DirectWildVarAtoms);
+  CS.getChkCG().getSuccessors(CS.getWild(), DirectWildVarAtoms);
+
+  // Maps each atom to the set of atoms which depend on it through an
+  // implication constraint. These atoms would not be found through a BFS
+  // because and explicit edge does not exist, but wildness can still flow
+  // between them when the implication is fired.
```

Comment:
But won't you know that it has, or has not, fired by the time you care about the statistics? The implication constraints get added if they are used, during solving, right?

Reply:
This is after solving, so I know which variables are WILD. I'm only looking at these variables for the root cause analysis.  Since they are WILD, if there was an implication constraint, it must have fired.

Reply:
Sorry, I asked the wrong question. Why do we care about implication constraints being the ones that caused the problem, specifically? If they have been added to the graph, why isn't that enough? Is it about "blame" ?

Reply:
When the constraint fires a new edge WILD->VAR is added to the graph. I don't want this edge to show up as root cause. Instead, whatever constraint caused the  implication to fire should be given all the atoms affected by the implication.

Reply:
Aha. Then I would change the second half of your comment above to say something to the effect of: We remember implication edges to avoid misattributing them in root causes. Why? WILDness introduced by implication constraints is, by definition, derived from the WILDness of some other non-implication edge.